src/UglyToad.PdfPig/DocumentLayoutAnalysis/RecursiveXYCut.cs

```diff
@@ -0,0 +1,408 @@
+            }
+        }
+
+        private void RecursiveCount(IEnumerable<XYNode> children, ref int count)
+        {
+            if (children.Count() == 0) return;
+            foreach (XYNode node in children.Where(x => x.IsLeaf))
+            {
+                count += node.CountWords();
+            }
+
+            foreach (XYNode node in children.Where(x => !x.IsLeaf))
+            {
+                RecursiveCount(node.Children, ref count);
+            }
+        }
+
+        private void RecursiveGetLeafs(IEnumerable<XYNode> children, ref List<XYLeaf> leafs, int level)
+        {
+            if (children.Count() == 0) return;
+            bool isVerticalCut = level % 2 == 0;
+
+            foreach (XYLeaf node in children.Where(x => x.IsLeaf))
+            {
+                leafs.Add(node);
+            }
+
+            level++;
+
+            IEnumerable<XYNode> notLeafs = children.Where(x => !x.IsLeaf);
+
+            if (isVerticalCut)
+            {
+                notLeafs = notLeafs.OrderBy(x => x.BoundingBox.Left).ToList();
+            }
+            else
+            {
+                notLeafs = notLeafs.OrderByDescending(x => x.BoundingBox.Top).ToList();
+            }
+
+            foreach (XYNode node in notLeafs)
+            {
+                RecursiveGetLeafs(node.Children, ref leafs, level);
+            }
+        }
+
+        public override string ToString()
+        {
+            return (IsLeaf ? "Leaf" : "Node");
+        }
+    }
+
+    /// <summary>
+    /// A Leaf node used in the <see cref="RecursiveXYCut"/> algorithm, i.e. a block.
+    /// </summary>
+    public class XYLeaf : XYNode
+    {
+        /// <summary>
+        /// Returns true if this node is a leaf, false otherwise.
+        /// </summary>
+        public override bool IsLeaf => true;
+
+        /// <summary>
+        /// The words in the leaf.
+        /// </summary>
+        public Word[] Words { get; set; }
+
+        /// <summary>
+        /// The number of words in the leaf.
+        /// </summary>
+        public override int CountWords() => Words == null ? 0 : Words.Length;
+
+        /// <summary>
+        /// Returns null as a leaf doesn't have leafs.
+        /// </summary>
+        public override List<XYLeaf> GetLeafs()
+        {
+            return null;
+        }
+
+        /// <summary>
+        /// Gets the lines of the leaf.
+        /// </summary>
+        public TextLine[] GetLines()
+        {
+            var groupedWords = Words.GroupBy(x => x.BoundingBox.Bottom).ToDictionary(x => x.Key, x => x.ToList());
+            return groupedWords.OrderByDescending(x => x.Key).Select(ws => new TextLine(ws.Value)).ToArray();
+        }
+
+        /// <summary>
+        /// Create a new <see cref="XYLeaf"/>.
+        /// </summary>
+        /// <param name="words">The words contained in the leaf.</param>
+        public XYLeaf(params Word[] words) : this(words == null ? null : words.ToList())
```

Comment:
I think it's better to leave it like this so that it does not throw here. The check will be added in `public XYLeaf(IEnumerable<Word> words) : base(null)`